From c16a44b6627a6b4c1cb178f4c2127f21a53344ec Mon Sep 17 00:00:00 2001
From: Vojtech Trefny <vtrefny@redhat.com>
Date: Mon, 3 Mar 2025 12:33:34 +0100
Subject: [PATCH] iscsi: Use node.startup=onboot option for Login

Resolves: RHEL-53719
---
 blivet/iscsi.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/blivet/iscsi.py b/blivet/iscsi.py
index 95674665b..f66c38934 100644
--- a/blivet/iscsi.py
+++ b/blivet/iscsi.py
@@ -278,7 +278,7 @@ def _login(self, node_info, extra=None):
 
         if extra is None:
             extra = dict()
-        extra["node.startup"] = GLib.Variant("s", "automatic")
+        extra["node.startup"] = GLib.Variant("s", "onboot")
         extra["node.session.auth.chap_algs"] = GLib.Variant("s", "SHA1,MD5")
 
         args = GLib.Variant("(sisisa{sv})", node_info.conn_info + (extra,))
